### PR TITLE
Multi target regeression

### PR DESCRIPTION
**Pull request by nickwp**
This PR is mainly to enable training regression of multiple quantities (targets) simultaneously, e.g. a single model to reconstruct both position and direction.

To allow this to work smoothly, some other changes and bug fixes were necessary, and some other things modified to make things more consistent.

Summary of changes:
- Ability to provide a list of `target_key` values to the engine instead of just a single value. This is renamed from the old `truth_key` config parameter, as it defines what the target values of the regression are by specifying which element(s) of the dictionary provided by the dataset.
- Regression output saves the true and predicted values for each target
- Added print-out of regression metrics during training (different for each target type: e.g. mean 3D distance from true to predicted vertex position, mean angle between true and predicted direction, mean fractional residual of energy prediction). This is easier to interpret than raw loss values.
- Reworked target scaling to accommodate multiple targets (important hyperparameters to balance the training loss across different targets). The config now takes a dictionary of targets and their corresponding scale factors and offsets, similarly to the existing channel scaling for scaling charge and time values, for example.
- Reworked the channel scaling config to have separate entries for the scale factor and offsets, so you don't need both and don't use a list with arbitrary ordering but use more explicitly named config parameters
- Bugfix: `drop_last` added to the data loader config, which should be used for validation to ensure we don't validate on an incomplete batch.
- Bugfix: force multi-GPU syncronisation between tasks to prevent rare cases where e.g. in the very last batch of training, a new best state is found at validation, but the "best" model is loaded by one GPU before another GPU finished training and saved the new best state
- Bugfix: correct calculation of average evaluation metrics for uneven batches

### CODE REVIEW

**Comment by nickwp** [comment]
Tested this with WCTE. After some basic hyperparameter tuning, achieved improved position and direction regression using a single model than the old separate models.

Probably hyper-parameter tuning of the old approach could also reach the same performance, but at least this validates the multi-target regression and shows it is potentially a useful feature.

Tuning the target scaling parameters, which affects the balance between the different targets' contributions to the loss, and the Huber loss delta value, which will depend on the target scale, does seem to have a noticeable impact on performance. So properly tuning this would be useful.

In the plots, the "multi" lines are for different combinations of hyperparameters.

Untitled
Untitled

**Comment by nickwp** [comment]
Merging this now since the results suggest this is working and it has been waiting a while...
